pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,231 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package telemetry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/about"
+	apmv1 "github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1"
+	beatv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/beat/v1beta1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	entv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/enterprisesearch/v1beta1"
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/kibana"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/ghodss/yaml"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	updateInterval = 1 * time.Hour
+	resourceCount  = "resource_count"
+	podCount       = "pod_count"
+)
+
+var log = logf.Log.WithName("usage")
+
+type getStatsFn func(k8s.Client, []string) (string, interface{}, error)
+
+func NewReporter(info about.OperatorInfo, client client.Client, managedNamespaces []string) Reporter {
+	if len(managedNamespaces) == 0 {
+		// treat no managed namespaces as managing all namespaces, ie. set empty string for namespace filtering
+		managedNamespaces = append(managedNamespaces, "")
+	}
+
+	return Reporter{
+		operatorInfo:      info,
+		client:            k8s.WrapClient(client),
+		managedNamespaces: managedNamespaces,
+	}
+}
+
+type Reporter struct {
+	operatorInfo      about.OperatorInfo
+	client            k8s.Client
+	managedNamespaces []string
+}
+
+func (r *Reporter) Start() {
+	ticker := time.NewTicker(updateInterval)
+	for range ticker.C {
+		r.report()
+	}
+}
+
+func marshalTelemetry(info about.OperatorInfo, stats map[string]interface{}) ([]byte, error) {
+	return yaml.Marshal(struct {
+		ECK struct {
+			about.OperatorInfo
+			Stats map[string]interface{} `json:"stats"`
+		} `json:"eck"`
+	}{
+		ECK: struct {
```

Comment:
anonymous structs are a bit verbose, but I guess your rationale is that it is used only once?

Reply:
Yes. We can create types if we use it in more than a single place.